File: src/main/kotlin/commands/RemoveVersion.kt
package commands

import Environment
import Errors
import Errors.doesNotExistError
import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.ProgramResult
import com.github.ajalt.clikt.core.terminal
import com.github.ajalt.clikt.parameters.options.check
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.mordant.animation.progressAnimation
import com.github.ajalt.mordant.rendering.TextColors
import com.github.ajalt.mordant.terminal.ConversionResult
import com.github.ajalt.mordant.terminal.Terminal
import data.shared.PackageIdentifier
import data.shared.PackageVersion
import github.GitHubImpl
import github.GitHubUtils
import io.ExitCode
import io.Prompts
import io.menu.yesNoMenu
import kotlinx.coroutines.runBlocking
import org.kohsuke.github.GHContent
import org.kohsuke.github.GitHub
import token.Token
import token.TokenStore
import utils.versionStringComparator

class RemoveVersion : CliktCommand(
    help = """
        Removes a pre-existing package version in winget-pkgs
        
        To remove an entire package, all versions of that package must be removed
    """.trimIndent(),
    name = "remove"
) {
    private val packageIdentifierParam: String? by option(
        "-i", "--id", "--identifier", "--package-identifier",
        help = "Package identifier. Example: Publisher.Package"
    )

    private val packageVersionParam: String? by option(
        "-v", "--version", "--package-version",
        help = "Package version. Example: 1.2.3"
    )

    private val deletionReasonParam: String? by option("--reason", "--reason-for-deletion", "--deletion-reason")
        .check(Errors.invalidLength(min = REASON_MIN_LENGTH, max = REASON_MAX_LENGTH)) {
            it.length in REASON_MIN_LENGTH..REASON_MAX_LENGTH
        }

    private val submit: Boolean by option(
        "-s", "--submit",
        help = "Automatically submits a pull request with the updated pull request"
    ).flag(default = false)

    private val token: String? by option(
        "-t", "--token", "--pat", "--personal-access-token",
        help = "GitHub personal access token with the public_repo scope",
        envvar = "GITHUB_TOKEN"
    ).check("The token is invalid or has expired") { GitHub.connectUsingOAuth(it).isCredentialValid }

    private lateinit var packageIdentifier: String
    private lateinit var packageVersion: String

    override fun run(): Unit = runBlocking {
        token?.let { TokenStore.useTokenParameter(it) }
        if (TokenStore.token == null) prompt(Token).also { TokenStore.putToken(it) }
        warning("Packages should only be removed when necessary.")
        echo()
        packageIdentifier = prompt(PackageIdentifier, parameter = packageIdentifierParam)
        if (!TokenStore.isTokenValid.await()) TokenStore.invalidTokenPrompt(terminal)
        val allVersions = GitHubUtils.getAllVersions(GitHubImpl.microsoftWinGetPkgs, packageIdentifier)?.also {
            info("Found $packageIdentifier in the winget-pkgs repository")
            it.maxWithOrNull(versionStringComparator)?.let { latestVersion ->
                info("Found latest version: $latestVersion")
            }
        }
        packageVersion = prompt(PackageVersion, parameter = packageVersionParam)
        GitHubImpl.microsoftWinGetPkgs.getDirectoryContent(GitHubUtils.getPackagePath(packageIdentifier))
            ?.find { it.name == packageVersion }
            ?: throw doesNotExistError(packageIdentifier, packageVersion, theme = theme)
        val deletionReason = deletionReasonParam ?: terminal.promptForDeletionReason()
        val shouldRemoveManifest = if (submit || Environment.isCI) {
            true
        } else {
            info("Would you like to make a pull request to remove $packageIdentifier $packageVersion?")
            terminal.yesNoMenu(default = true).prompt()
        }
        if (!shouldRemoveManifest) return@runBlocking
        echo()
        val forkRepository = GitHubImpl.getWingetPkgsFork(terminal)
        val ref = GitHubImpl.createBranchFromUpstreamDefaultBranch(
            winGetPkgsFork = forkRepository,
            packageIdentifier = packageIdentifier,
            packageVersion = packageVersion
        ) ?: return@runBlocking
        val directoryContent: MutableList<GHContent> = forkRepository
            .getDirectoryContent(GitHubUtils.getPackageVersionsPath(packageIdentifier, packageVersion), ref.ref)
        val progress = terminal.progressAnimation {
            text("Deleting files")
            percentage()
            progressBar()
            completed()
        }
        progress.start()
        directoryContent.forEachIndexed { index, ghContent ->
            ghContent.delete("Remove: ${ghContent.name}", ref.ref)
            progress.update(index.inc().toLong(), directoryContent.size.toLong())
        }
        progress.clear()
        GitHubImpl.microsoftWinGetPkgs.createPullRequest(
            "Remove: $packageIdentifier version $packageVersion",
            "${GitHubImpl.forkOwner}:${ref.ref}",
            GitHubImpl.microsoftWinGetPkgs.defaultBranch,
            "## $deletionReason"
        ).also { success("Pull request created: ${it.htmlUrl}") }
    }

    private fun Terminal.promptForDeletionReason(): String {
        echo(TextColors.brightGreen("${Prompts.REQUIRED} Give a reason for removing this manifest"))
        return prompt("Reason") {
            if (it.length < REASON_MIN_LENGTH || it.length > REASON_MAX_LENGTH) {
                ConversionResult.Invalid(Errors.invalidLength(min = REASON_MIN_LENGTH, max = REASON_MAX_LENGTH))
            } else {
                ConversionResult.Valid(it)
            }
        } ?: throw ProgramResult(ExitCode.CTRLC)
    }

    companion object {
        const val REASON_MIN_LENGTH = 4
        const val REASON_MAX_LENGTH = 1000
    }
}

File: src/commands/remove_version.rs
use std::num::NonZeroU32;

use anstream::println;
use clap::Parser;
use color_eyre::eyre::{Result, bail};
use futures_util::TryFutureExt;
use inquire::{
    Text,
    validator::{MaxLengthValidator, MinLengthValidator},
};
use owo_colors::OwoColorize;
use tokio::try_join;
use winget_types::{PackageIdentifier, PackageVersion};

use crate::{
    credential::handle_token,
    github::github_client::{GitHub, WINGET_PKGS_FULL_NAME},
    prompts::{handle_inquire_error, text::confirm_prompt},
};

/// Remove a version from winget-pkgs
///
/// To remove a package, all versions of that package must be removed
#[derive(Parser)]
pub struct RemoveVersion {
    /// The package's unique identifier
    #[arg()]
    package_identifier: PackageIdentifier,

    /// The package's version
    #[arg(short = 'v', long = "version")]
    package_version: PackageVersion,

    #[arg(short = 'r', long = "reason")]
    deletion_reason: Option<String>,

    /// List of issues that removing this version would resolve
    #[arg(long)]
    resolves: Option<Vec<NonZeroU32>>,

    #[arg(short, long)]
    submit: bool,

    /// Don't show the package removal warning
    #[arg(long)]
    no_warning: bool,

    /// Open pull request link automatically
    #[arg(long, env = "OPEN_PR")]
    open_pr: bool,

    /// GitHub personal access token with the `public_repo` scope
    #[arg(short, long, env = "GITHUB_TOKEN")]
    token: Option<String>,
}

impl RemoveVersion {
    const MIN_REASON_LENGTH: usize = 4;
    const MAX_REASON_LENGTH: usize = 1000;

    pub async fn run(self) -> Result<()> {
        let token = handle_token(self.token.as_deref()).await?;
        if !self.no_warning {
            println!(
                "{}",
                "Packages should only be removed when necessary".yellow()
            );
        }
        let github = GitHub::new(&token)?;

        let (fork, winget_pkgs, versions) = try_join!(
            github
                .get_username()
                .and_then(|current_user| github.get_winget_pkgs().owner(current_user).send()),
            github.get_winget_pkgs().send(),
            github.get_versions(&self.package_identifier)
        )?;

        if !versions.contains(&self.package_version) {
            bail!(
                "{} version {} does not exist in {WINGET_PKGS_FULL_NAME}",
                self.package_identifier,
                self.package_version,
            );
        }

        let latest_version = versions.last().unwrap_or_else(|| unreachable!());
        println!(
            "Latest version of {}: {latest_version}",
            &self.package_identifier
        );
        let deletion_reason = match self.deletion_reason {
            Some(reason) => reason,
            None => Text::new(&format!(
                "Give a reason for removing {} version {}",
                &self.package_identifier, &self.package_version
            ))
            .with_validator(MinLengthValidator::new(Self::MIN_REASON_LENGTH))
            .with_validator(MaxLengthValidator::new(Self::MAX_REASON_LENGTH))
            .prompt()
            .map_err(handle_inquire_error)?,
        };
        let should_remove_manifest = self.submit
            || confirm_prompt(&format!(
                "Would you like to make a pull request to remove {} {}?",
                self.package_identifier, self.package_version
            ))?;

        if !should_remove_manifest {
            return Ok(());
        }

        let pull_request_url = github
            .remove_version()
            .identifier(&self.package_identifier)
            .version(&self.package_version)
            .reason(deletion_reason)
            .fork(&fork)
            .winget_pkgs(&winget_pkgs)
            .maybe_issue_resolves(self.resolves)
            .send()
            .await?;

        if self.open_pr {
            open::that(pull_request_url.as_str())?;
        }

        Ok(())
    }
}
